### PR TITLE
feat: update flight-sql to latest specs

### LINKED ISSUE #4249: Update flight-sql implementation to latest specs
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The proto definitions for the flight sql spec have seen some updates not yet reflected in the `arrow-flight` crate.

**Describe the solution you'd like**

Update and build latest proto files and update current implementations to reflect changes.

**Describe alternatives you've considered**

Not keeping up to date :).

**Additional context**

### PR DESCRIPTION
**Pull request by roeap**
# Which issue does this PR close?

Closes #4249

# Rationale for this change
 
arrow-rs should keep up-to-date with the reference specs from apache/arrow.

# What changes are included in this PR?

* generate types based on latest proto files from apache/arrow
* update flight-sql server / client implementations

# Are there any user-facing changes?

New actions and statements are added. The flight sql server trait has corresponding handlers added.

### CODE REVIEW

**Review by tustvold — COMMENTED** [review]
I left some comments, but this probably wants a review by someone more familiar with FlightSQL, perhaps @alamb or @avantgardnerio

File: arrow-flight/src/sql/server.rs [hunk 1]
```diff
@@ -529,9 +656,108 @@ where
                         "Unable to unpack ActionClosePreparedStatementRequest.",
                     )
                 })?;
-            self.do_action_close_prepared_statement(cmd, request).await;
+            self.do_action_close_prepared_statement(cmd, request)
+                .await?;
             return Ok(Response::new(Box::pin(futures::stream::empty())));
         }
+        if request.get_ref().r#type == CREATE_PREPARED_SUBSTRAIT_PLAN {
```
[the remark this answers is not in the document]
Comment (roeap) [reply arrow-flight/src/sql/server.rs:663]: just matched what was there ... but yes, adding some `else`s makes much more sene :)

[the remark this answers is not in the document]
Comment (roeap) [reply arrow-flight/src/sql/server.rs]: updated to `eles if` ... i guess since we return in every branch, it's not too different though?

**Comment by alamb** [comment]
>  @alamb - playing around with implementing a flight sql server I read a bit through the iOx implementation. Over there I found that the structs around sql infos (SqlInfoValue, SqlInfoList, SqlInfoUnionBuilder) are quite useful and probably non trivial to figure out for users. Do you think this is something that could be hoisted upstream?

Yes, absolutely we could hoist this upstream. 

I filed these tickets to do so:
* `CommandGetSqlInfo`: https://github.com/apache/arrow-rs/issues/4256
* `CommandGetXdbcTypeInfo`: https://github.com/apache/arrow-rs/issues/4257

There are several other things that could be hoisted upstream (like the basic implementations for `GetTables` and `GetSchemas` both of which need basic filtering). Do you think those are worth pulling upstream too?

**Comment by alamb** [comment]
I will review this PR carefully later today or tomorrow. Thank you @roeap

File: arrow-flight/examples/flight_sql_server.rs [hunk 2]
```diff
@@ -486,10 +512,58 @@ impl FlightSqlService for FlightSqlServiceImpl {
...
+    ) -> Result<(), Status> {
         unimplemented!("Implement do_action_close_prepared_statement")
     }
 
+    async fn do_action_create_prepared_substrait_plan(
+        &self,
+        _query: ActionCreatePreparedSubstraitPlanRequest,
+        _request: Request<Action>,
+    ) -> Result<ActionCreatePreparedStatementResult, Status> {
+        unimplemented!("Implement do_action_create_prepared_substrait_plan")
```
[the remark this answers is not in the document]
Comment (roeap) [reply arrow-flight/examples/flight_sql_server.rs]: There is a status macro defined, which also reports line numbers.

```rs
macro_rules! status {
    ($desc:expr, $err:expr) => {
        Status::internal(format!("{}: {} at {}:{}", $desc, $err, file!(), line!()))
    };
}
```

Maybe we could add another macro that uses `Status::unimplemented` to communicate the right status and have the line info available?

**Comment by roeap** [comment]
> There are several other things that could be hoisted upstream (like the basic implementations for GetTables and GetSchemas both of which need basic filtering). Do you think those are worth pulling upstream too?

Definitely - should they however go here, or would they need to be in the datafusion example server and include extracting those from the session context?

**Review by avantgardnerio — APPROVED** [review]
So in summary:
- FlightSQL now supports transactions IDs (very cool)
- it can also execute subtrait plans (super neat)
- better docs

LGTM

**Comment by alamb** [comment]
> Definitely - should they however go here, or would they need to be in the datafusion example server and include extracting those from the session context?

I think they can be implemented directly with the arrow kernels without datafusion -- the comparisons needed are pretty simple (string matching) and we did this in IOx to avoid the overhead of creating a query plan.

> Also, I'd be happy to take on these issues and move the code to arrow-flight.

Thank you @roeap  -- that is awesome. ❤️  I will plan to file the other tickets tomorrow and would love to help design / review the PRs.

**Review by alamb — APPROVED** [review]
Thank you @roeap  -- this looks great. 

I double checked the proto files matched at https://github.com/apache/arrow/commit/fbe5f641d327ee81db00ce5f056940a69f4d8603

```shell
(arrow_dev) alamb@MacBook-Pro-8 arrow-rs % diff -du -w ./format/Flight.proto /Users/alamb/Software/arrow/format/Flight.proto
diff -du -w ./format/Flight.proto /Users/alamb/Software/arrow/format/Flight.proto
(arrow_dev) alamb@MacBook-Pro-8 arrow-rs % diff -du -w ./format/FlightSql.proto  /Users/alamb/Software/arrow/format/FlightSql.proto
diff -du -w ./format/FlightSql.proto  /Users/alamb/Software/arrow/format/FlightSql.proto
(arrow_dev) alamb@MacBook-Pro-8 arrow-rs % 
```

**Pull request merged**